staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,602 @@
+// a queue and eventually dispatches the request from that queue.
+// Dispatching means to return with `quiescent==false` and
+// `execute==true`.  In one unhappy case the request is
+// immediately rebuffed with `quiescent==true` (which tells the
+// filter that there has been a timing splinter and the filter
+// re-calcuates the priority level to use); in all other cases
+// `quiescent` will be returned `false` (even if the system is
+// quiescent by then).  In the non-quiescent unhappy cases the
+// request is eventually rejected, which means to return with
+// `execute=false`.  In the happy case the caller is required to
+// invoke the returned `afterExecution` after the request is done
+// executing.  The hash value and hand size are used to do the
+// shuffle sharding.
+func (qs *queueSetImpl) Wait(hashValue uint64) (quiescent, execute bool, afterExecution func()) {
+	// TODO(aaron-prindle) verify locking is correct
+
+	qs.lockAndSyncTime()
+	// TODO(aaron-prindle) verify and test quiescent
+	// A call to Wait while the system is quiescent will be rebuffed by
+	// returning `quiescent=true`.
+	if qs.emptyHandler != nil {
+		qs.lock.Unlock()
+		return true, false, func() {}
+	}
+
+	// ========================================================================
+	// Step 1:
+	// 1) Start with shuffle sharding, to pick a queue.
+	// 2) Reject old requests that have been waiting too long
+	// 3) Reject current request if there is not enough concurrency shares and
+	// we are at max queue length
+	// 4) If not rejected, create a packet and enqueue
+	pkt := qs.timeoutOldRequestsAndRejectOrEnqueue(hashValue)
+	// pkt == nil means that the request was rejected - no remaining
+	// concurrency shares and at max queue length already
+	if pkt == nil {
+		qs.lock.Unlock()
+		metrics.AddReject(qs.name, "queue-full")
+		return false, false, func() {}
+	}
+	// ========================================================================
+
+	// ------------------------------------------------------------------------
+	// Step 2:
+	// 1) The next step is to invoke the method that dequeues as much as possible.
+
+	// This method runs a loop, as long as there
+	// are non-empty queues and the number currently executing is less than the
+	// assured concurrency value.  The body of the loop uses the fair queuing
+	// technique to pick a queue, dequeue the request at the head of that
+	// queue, increment the count of the number executing, and send true to
+	// the request's channel.
+	qs.dequeueWithChannelAsMuchAsPossible()
+	// ------------------------------------------------------------------------
+	qs.lock.Unlock()
+
+	// ************************************************************************
+	// Step 3:
+	// After that method finishes its loop and returns, the final step in Wait
+	// is to `select` (wait) on a message from the enqueud packet's channel
+	// and return appropriately.
+	qs.counter.Add(-1)
+
+	select {
```

Comment:
To plumb cancellation into here, we need to add a case to this `select`.  The new case fires when the cancellation channel is closed.  The code to run in that case will need to invoke a new entry method that ensures the Request is not in the Queue.  That will require at least adding a status bit into the `Request` type to indicate whether it is in its Queue.  If the Request _is_ in the Queue then it will need to be removed.  We _could_ code that as a simple linear search of the Queue's slice of `*Request` and then splicing out the sought entry if it is found.  That costs O(len(queue)).  I think this should be rare enough and queue lengths should be kept short enough that this is an acceptable cost for our first release.  Later we might switch a Queue to using a doubly-linked list instead of a slice.

Reply:
Introducing a channel will also require care regarding the go routine counting.  I think we can do it, remembering that we only need to get it right for tests.

Reply:
Plumbed through using a full context.Context and the suggested simple approach:
```
	case <-ctx.Done():
		klog.V(5).Infof("request timed out\n")
		for i := range req.Queue.Requests {
			if req == req.Queue.Requests[i] {
				removeRequest(req.Queue.Requests, i)
				break
			}
		}
		return false, false, func() {}
```

I don't think there are additional locking/race considerations for this piece.  If we are in the timeout case for the switch statement, the request will either still be in the queue or will have been dequeued but not executing.

Reply:
The timeout handling code _does_ need to hold the QueueSet lock, otherwise there could be confusion with a concurrent goroutine dispatching the request.

Reply:
I've updated the code to hold a lock for all access of `qs`